lint: Add commit/PR body linter for epic and issue refs

nickvigilante opened the pull request:
Before: Commits and PRs would inconsistently contain references to issues.
Sometimes the author would forget to add it. Some authors might not
think to add them. And there were no epic references provided.

Why: Adding issue and epic references to PRs and commits provides
traceability and context for the revenue teams, product, management and
engineers.

Now: A new GitHub Action workflow checks for references to issues the PR
closes or informs and epics the PR is part of. It looks for references
in the PR body and in commit messages and, if it doesn't find what it
expects, it fails the check.

This PR is a continuation of #77654.

Fixes #77376

Release note: None

Release justification: Adding linter for better epic tracking

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87128)
<!-- Reviewable:end -->

nickvigilante commented:
@jlinder This is ready for final review. My last Bazel run reached the 90 minute timeout.

Commit:
lint: Add commit/PR body linter for epic and issue refs

Before: Commits and PRs would inconsistently contain references to issues.
Sometimes the author would forget to add it. Some authors might not
think to add them. And there were no epic references provided.

Why: Adding issue and epic references to PRs and commits provides
traceability and context for the revenue teams, product, management and
engineers.

Now: A new GitHub Action workflow checks for references to issues the PR
closes or informs and epics the PR is part of. It looks for references
in the PR body and in commit messages and, if it doesn't find what it
expects, it fails the check.

Release note: None

Release justification: Adding linter for epic tracking for engineering,
docs, and internal teams.

jlinder commented on pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go:
```diff
@@ -0,0 +1,319 @@
+)
+
+var (
+	fixIssueRefRE        = regexp.MustCompile(`(?im)(?i:close[sd]?|fix(?:e[sd])?|resolve[sd]?):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+)|([A-Z]+-\d+))[,.;]?(?:[ \t\n\r]+|$))+`)
+	informIssueRefRE     = regexp.MustCompile(`(?im)(?:see also|informs):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+)|([A-Z]+-\d+))[,.;]?(?:[ \t]+|$))+`)
+	epicRefRE            = regexp.MustCompile(`(?im)epic:?\s+(?:([A-Z]+-[0-9]+)[,.;]?(?:[ \t]+|$))+`)
+	epicNoneRE           = regexp.MustCompile(`(?im)epic:?\s+(?:(none)[,.;]?(?:[ \t]+|$))+`)
```
Good find on adding the `\n\r`. Let's fix it for the other three regexes too. Also, I think the `+` after the `[ \t\n\r]` can be removed. The regexes only need to verify that there is at least one separating character there.

```suggestion
	fixIssueRefRE        = regexp.MustCompile(`(?im)(?i:close[sd]?|fix(?:e[sd])?|resolve[sd]?):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+)|([A-Z]+-\d+))[,.;]?(?:[ \t\n\r]|$))+`)
	informIssueRefRE     = regexp.MustCompile(`(?im)(?:see also|informs):?\s+(?:(?:(#\d+)|([\w.-]+[/][\w.-]+#\d+)|([A-Z]+-\d+))[,.;]?(?:[ \t\n\r]|$))+`)
	epicRefRE            = regexp.MustCompile(`(?im)epic:?\s+(?:([A-Z]+-[0-9]+)[,.;]?(?:[ \t\n\r]|$))+`)
	epicNoneRE           = regexp.MustCompile(`(?im)epic:?\s+(?:(none)[,.;]?(?:[ \t\n\r]|$))+`)
```

nickvigilante replied:
I added `\n\r` to each of the other RegExes, but I kept the `+`, because tests were failing without it.

jlinder commented on pkg/cmd/lint-epic-issue-refs/lint_epic_issue_refs.go:
```diff
@@ -0,0 +1,319 @@
+		checkMultipleEpicsInCommits(commitInfos, params)
+
+	if !passedAllTests {
+		fmt.Println("See also: https://wiki.crdb.io/wiki/spaces/CRDB/pages/2009039063/")
```
```suggestion
        fmt.Println("For more information about issue and epic references, see: https://wiki.crdb.io/wiki/spaces/CRDB/pages/2009039063/")
```

Suggesting just the text change. The GitHub UI won't allow entry of tabs.

nickvigilante replied:
Done.

jlinder reviewed (approved):
Had one suggestion. Good either way!

nickvigilante commented:
bors r+
